src/client/opamClient.ml

```diff
@@ -184,11 +184,21 @@ let compute_upgrade_t
              then (n, None)
              else
              let atom = (n, Some (`Gt, nv.version)) in
+             let has_hidden_version nv =
+               OpamFile.OPAM.has_flag Pkgflag_HiddenVersion
+                 (OpamSwitchState.opam t nv)
+             in
+             let hidden_version_already_installed pkg =
+               OpamPackage.Set.exists (fun nv ->
+                   OpamPackage.Name.equal (OpamPackage.name nv) pkg &&
+                   has_hidden_version nv
+                 ) t.installed
+             in
```

Comment:
If it is used in 2 places, it is possible to add a function `OpamSwitchState.has_hidden_version_and_installed` for ex. Like that if a change need to be done, it would be only in one place